### PR TITLE
(Experimental) Make wifi cmd RESTful

### PR DESCRIPTION
**Pull request by thienhoang23**
* As a proof of concept, write an http server that will start when wifi cmd is run, thus allowing user to interact with the cmd over a RESTful interface.

* Since this is a proof of concept at the moment, stub data is used to prove that the concept works. 

* Test cases have been written and manual tests have been carried out to ensure backward compatability with previous version of wifi cmd.

### CODE REVIEW

File: cmds/wifi/wifi.go [hunk 1]
```diff
@@ -38,36 +38,62 @@ func init() {
 	}
 }
 
+func generateConfig(essid string, id string, pass string) (conf []byte, err error) {
+	switch {
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: func generateConfig(s ...string) 
switch len(s) {
case 1:
case 2: 
case 3:
default: 
etc.

File: cmds/wifi/wifi.go [hunk 2]
```diff
@@ -38,36 +38,62 @@ func init() {
...
 	)
 
 	flag.Parse()
 	a := flag.Args()
 
 	switch {
 	case len(a) == 3:
-		essid = a[0]
-		conf = []byte(fmt.Sprintf(eap, essid, a[2], a[1]))
+		conf, err = generateConfig(a[0], a[2], a[1])
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: it seems a little awkward to me to duplicate the logic of generateConfig here and in the function too.

you might just call generateConfig with
generateConfig(a...)

and then in the function itself, case on the len of args.

see https://play.golang.org/p/mWdAHqBBKjh

Comment (thienhoang23) [reply cmds/wifi/wifi.go]: I agree with what you said. My purpose for doing this initially was to promote code reuse becuase I noticed that after we get inputs from client, we still need to go through the same logic to generate the conf. I think I have a fix for this awkwardness coming up in the next commit.

File: cmds/wifi/wifi.go [hunk 3]
```diff
@@ -38,34 +38,59 @@ func init() {
...
+		conf = []byte(fmt.Sprintf(eap, a[0], a[2], a[1]))
+	case len(a) == 2:
+		conf, err = passphrase.Run(a[0], a[1])
+		if err != nil {
+			return nil, fmt.Errorf("essid: %v, pass: %v : %v", a[0], a[1], err)
+		}
+	case len(a) == 1:
+		conf = []byte(fmt.Sprintf(nopassphrase, a[0]))
+	default:
+		flag.Usage()
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: return an error
return nil, fmt.Errorf("generateConfig needs 0, 1, 2, or 3 args") or some similar thing.

File: cmds/wifi/wifi.go [hunk 4]
```diff
@@ -38,34 +38,59 @@ func init() {
 	}
 }
 
+func generateConfig(a []string) (conf []byte, err error) {
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: The convention I'm used to is a...string but I guess this is ok?

File: cmds/wifi/wifi.go [hunk 5]
```diff
@@ -38,34 +38,59 @@ func init() {
...
-		conf = o
-	case len(a) == 1:
-		essid = a[0]
-		conf = []byte(fmt.Sprintf(nopassphrase, essid))
-	default:
-		flag.Usage()
-		os.Exit(1)
+		if msg.id != "" {
+			a = append(a, msg.id)
+		}
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: Question: why can't the msg just contain a slice of strings, then you don't have to do this append stuff.

File: cmds/wifi/wifi.go [hunk 6]
```diff
@@ -38,34 +38,59 @@ func init() {
...
+		ServiceToServerChan <- stubMsg
+		_ = <-ServerToServiceChan // (Experimental) So we can see the result of the page load
+	}
+
+	conf, err = generateConfig(a)
+	if err != nil {
+		log.Fatalf("error: %v", err)
 	}
 
 	if err := ioutil.WriteFile("/tmp/wifi.conf", conf, 0444); err != nil {
```
Comment (rminnich) [inline cmds/wifi/wifi.go]: so we only see the wpa_supplicant output if exec.Command got the error? What is the supplicant exits normally? Shouldn't we see its output if it is not empty?

File: cmds/wifi/wifi.go (hunk 3 above)
Comment (thienhoang23) [reply cmds/wifi/wifi.go]: Fixed

File: cmds/wifi/wifi.go (hunk 4 above)
Comment (thienhoang23) [reply cmds/wifi/wifi.go]: Fixed

File: cmds/wifi/wifi.go (hunk 5 above)
Comment (thienhoang23) [reply cmds/wifi/wifi.go]: Fixed

File: cmds/wifi/wifi.go (hunk 6 above)
Comment (thienhoang23) [reply cmds/wifi/wifi.go]: Sorry, I don't quite understand this comment. The only error that can come out from generateConfig is if wpa_passphrase has an error or if invalid argument is passed in. Any of those two will cause the to exit which is similar to what we had before. Those lines are just standard error checking.

**Pull request merged**